# Week 2 Homework: Draft of Data Preprocessing of Machine Learning Solution Project Dataset


---


Data preprocessing ensures that the data used to develop machine learning solutions is clean, consistent, and relevant. For the next two weeks, you will preprocess the FinMark Corporation datasets. The main objective of this activity is to exercise your familiarity and skills in data preprocessing using Python.

<br>

To carry out this activity, follow the instructions below:

1. Access the company website here. Review the problem statement and download the provided datasets.

2. Load the downloaded datasets in Jupyter Notebooks and preprocess them using Pandas:

    a. Check for duplicate rows and remove them.

    b. Identify columns with missing values and decide how to handle them.

    c. Ensure columns have the appropriate data types.

3. Once completed, save the preprocessed datasets and prepare to discuss the output with your team.

To access files stored in Google Drive from Google Colab, it's necessary to connect Google Drive to the Colab environment. This allows direct access to files stored in Drive. Follow these steps to mount Google Drive.

In [1]:
# Import the module to interact with Google Drive in Colab
from google.colab import drive

# Mount Google Drive to access files stored in your drive from Colab
drive.mount('/content/drive')

from google.colab import auth
auth.authenticate_user()

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## **Import Pandas and Read csv**

To work with data from three CSV files in Python using pandas, first, import the pandas library. Then, load each CSV file into separate DataFrame variables df1, df2, and df3 by specifying the file paths.

In [2]:
# Import the pandas library
import pandas as pd

# Reading the CSV file into Pandas DataFrame
#Customer Segments
df1 = pd.read_csv('/content/drive/MyDrive/3rd year 2024-25/MO-IT125 - Math for Machine Learning: Linear Algebra/FinMark csv/customer_demographics_contaminated.csv')
#Customer Transactions
df2 = pd.read_csv('/content/drive/MyDrive/3rd year 2024-25/MO-IT125 - Math for Machine Learning: Linear Algebra/FinMark csv/customer_transactions_contaminated.csv')
#Social Media Interactions
df3 = pd.read_csv('/content/drive/MyDrive/3rd year 2024-25/MO-IT125 - Math for Machine Learning: Linear Algebra/FinMark csv/social_media_interactions_contaminated.csv')


## **`info()` function**

After loading data into pandas DataFrames (df1, df2, df3), it's helpful to use the `info()` function. This function provides an overview of the structure of each dataset, including:

The total number of entries (rows).
The column names and their data types (e.g., integers, floats, strings).
The number of non-null values in each column.
Using `info()` helps in understanding the dataset at a glance, identifying missing values, and ensuring that the data types are as expected for further analysis.

In [3]:
# Summary of the DataFrame, including column names, non-null counts, data types, and memory usage
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3200 entries, 0 to 3199
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   CustomerID   3200 non-null   object
 1   Age          2909 non-null   object
 2   Gender       3200 non-null   object
 3   Location     3200 non-null   object
 4   IncomeLevel  2897 non-null   object
 5   SignupDate   3200 non-null   object
dtypes: object(6)
memory usage: 150.1+ KB


In [4]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3200 entries, 0 to 3199
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   CustomerID       3200 non-null   object
 1   TransactionID    3200 non-null   object
 2   TransactionDate  3200 non-null   object
 3   Amount           2896 non-null   object
 4   ProductCategory  2901 non-null   object
 5   PaymentMethod    3200 non-null   object
dtypes: object(6)
memory usage: 150.1+ KB


In [5]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3200 entries, 0 to 3199
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   CustomerID       3200 non-null   object
 1   InteractionID    3200 non-null   object
 2   InteractionDate  3200 non-null   object
 3   Platform         2889 non-null   object
 4   InteractionType  3200 non-null   object
 5   Sentiment        2871 non-null   object
dtypes: object(6)
memory usage: 150.1+ KB


## **`describe()` function**

The `describe()` function provides a summary of the statistical properties of the numerical columns in a dataset. This includes:

* Count: The number of non-null entries in each column.
* Mean: The average value.
* Standard deviation (std): A measure of how spread out the values are.
* Minimum (min) and Maximum (max) values.
* Percentiles (25%, 50%, 75%): These help to understand the distribution of the data.

Using `describe()` gives a quick overview of the central tendency, dispersion, and distribution of the numerical data, which is useful for understanding its characteristics before performing further analysis.

In [6]:
# Generate summary statistics for numeric columns in df, including count, mean, standard deviation, min, max, and percentiles (25%, 50%, 75%)
df1.describe()


,CustomerID,Age,Gender,Location,IncomeLevel,SignupDate
count,3200,2909,3200,3200,2897,3200
unique,3000,58,2,2696,3,1544
top,1a5115af-2fa7-4198-b838-be2cdeccb768,26.0,Female,Port Michael,High,2024-06-08
freq,2,72,1610,6,1004,9


In [7]:
df2.describe()

,CustomerID,TransactionID,TransactionDate,Amount,ProductCategory,PaymentMethod
count,3200,3200,3200,2896,2901,3200
unique,1871,3000,808,2603,5,4
top,49f06ea9-4ae5-4ce5-98ce-d2c2f6f15cbd,5ff75116-0a50-4d04-80fb-31e5ccbb0769,2024-01-06,-100,Clothing,Debit Card
freq,7,2,13,39,598,809


In [8]:
df3.describe()

,CustomerID,InteractionID,InteractionDate,Platform,InteractionType,Sentiment
count,3200,3200,3200,2889,3200,2871
unique,1893,3000,450,3,3,5
top,fd35b6b4-ad1c-49cc-9c34-8c2d696f38e2,e77a8713-fdc7-4f71-85f8-9cd88cca0c42,2023-07-02,Instagram,Comment,Positive
freq,6,2,17,976,1079,962


## **Checking for duplicates**
After loading the datasets, it’s important to check for duplicate rows, as these can lead to inaccurate analysis.

This code loops through each DataFrame (`df1`, `df2`, and `df3`) and counts the number of duplicate rows using the `duplicated()` function. The result shows how many duplicate rows exist in each dataset, which can then be addressed if necessary.

In [9]:
# List of DataFrames
dataframes = [df1, df2, df3]

# Loop through each DataFrame to check for duplicates
for i, df in enumerate(dataframes, 1):
  num_duplicates = df.duplicated().sum()
  print(f"Number of duplicates in DataFrame {i}: {num_duplicates}")

Number of duplicates in DataFrame 1: 177
Number of duplicates in DataFrame 2: 185
Number of duplicates in DataFrame 3: 180


## **Removing duplicates - `drop_duplicates()` function**

The `drop_duplicates()` function removes any duplicate rows from each DataFrame. The `inplace=True` argument ensures that the changes are applied directly to the DataFrame.

Once duplicates are removed, the code checks each DataFrame again for any remaining duplicates using the `duplicated()` function. If duplicates are found, they are printed for further review; otherwise, a message is displayed confirming that no duplicates remain.

This two-step process ensures that the data is cleaned of duplicates.

In [10]:
# List of DataFrames
dataframes = [df1, df2, df3]

# Loop to remove duplicates from each DataFrame
for df in dataframes:
  df.drop_duplicates(inplace=True)

# Loop to check for duplicates in each DataFrame and print them if found
for df in dataframes:
  duplicates = df[df.duplicated()]
  if not duplicates.empty:
    print(f"Duplicates found:\n{duplicates}")
  else:
    print("No duplicates found.")

No duplicates found.
No duplicates found.
No duplicates found.


## **Checking for missing data**

This code loops through each DataFrame (`df1`, `df2`, `df3`), using the `isnull().sum()` function to calculate the number of missing values in each column. Only columns with missing values are displayed, along with the total count of missing entries for easier identification.

In [11]:
# List of DataFrames
dataframes = [df1, df2, df3]

# Loop through each DataFrame to check for missing values
for i, df in enumerate(dataframes, 1):
    print(f"Missing Values in DataFrame {i}:")
    # Calculate the sum of missing values for each column
    missing_values = df.isnull().sum()

    # Print columns with missing values
    print(missing_values[missing_values > 0])

    print("\n" + "="*50 + "\n")  # Separator for readability

Missing Values in DataFrame 1:
Age            276
IncomeLevel    285
dtype: int64


Missing Values in DataFrame 2:
Amount             283
ProductCategory    282
dtype: int64


Missing Values in DataFrame 3:
Platform     291
Sentiment    309
dtype: int64




## **Data Types**
Understanding the data types of each column in a dataset is crucial for performing accurate data analysis. The `dtypes` attribute in pandas provides the data type of each column in a DataFrame. This helps verify that the data types are appropriate for the operations intended, such as ensuring numerical columns are treated as floats and date columns are recognized as datetime objects.

The following code loops through each DataFrame (`df1`, `df2`, `df3`) and prints the data types for each column. A separator is added for readability between the outputs of different DataFrames.

In [12]:
# List of DataFrames
dataframes = [df1, df2, df3]

# Loop through each DataFrame to check data types
for i, df in enumerate(dataframes, 1):
    print(f"Data Types in DataFrame {i}:")

    # Print data types of each column
    print(df.dtypes)

    print("\n" + "="*50 + "\n")  # Separator for readability

Data Types in DataFrame 1:
CustomerID     object
Age            object
Gender         object
Location       object
IncomeLevel    object
SignupDate     object
dtype: object


Data Types in DataFrame 2:
CustomerID         object
TransactionID      object
TransactionDate    object
Amount             object
ProductCategory    object
PaymentMethod      object
dtype: object


Data Types in DataFrame 3:
CustomerID         object
InteractionID      object
InteractionDate    object
Platform           object
InteractionType    object
Sentiment          object
dtype: object




Pandas has recognized all columns as object, a generic type for strings or mixed data. However, each column should be converted to the correct data type for improved performance and clarity. Applying these changes will optimize the data and make it more efficient for analysis.

## **Date Formats**

In data analysis, having accurate and consistent date formats is essential. Date columns can significantly impact analyses, trends, and predictions. However, invalid or inconsistent date entries can lead to errors and misinterpretations.

The following script checks multiple DataFrames for the validity of their date columns. It converts date entries into a standard format and identifies any invalid dates. This process helps ensure data quality, making it easier to trust the results of any analysis.

This approach is valuable for data analysts and anyone working with time-related data. By detecting and addressing date format issues, we can lay a solid foundation for accurate analyses and better decision-making. The script will provide insights into any invalid date formats, allowing for quick fixes and improving overall data reliability.

In [13]:
# List of DataFrames to check and their respective date columns
dfs = [
    (df1, 'SignupDate'),
    (df2, 'TransactionDate'),
    (df3, 'InteractionDate')
]

# Loop over each DataFrame and its respective date column
for i, (df, date_column) in enumerate(dfs, 1):
    # Attempt to convert the date column to datetime, keeping original values intact
    converted_dates = pd.to_datetime(df[date_column], errors='coerce')

    # Find rows where the conversion failed (NaT values) and get the original date formats
    invalid_dates = df.loc[converted_dates.isna(), date_column]

    # Count the number of invalid date formats
    count_invalid = invalid_dates.shape[0]

    # Print the count and the original invalid date formats
    print(f"DataFrame {i} - Column '{date_column}':")
    print(f"Number of invalid date formats: {count_invalid}")

    if count_invalid > 0:
        print("Invalid date formats (original values):")
        print(invalid_dates.values)  # Print the original invalid date values
    else:
        print("No invalid date formats found.")

    print("="*50)  # Separator for better readability


DataFrame 1 - Column 'SignupDate':
Number of invalid date formats: 98
Invalid date formats (original values):
['30/01/2020' '12/01/2024' '10/01/2022' '14/01/2024' '07/07/2019'
 '25/11/2020' '10/06/2024' '18/03/2021' '27/06/2023' '24/04/2020'
 '03/04/2020' '04/02/2024' '22/04/2023' '06/06/2020' '16/02/2023'
 '11/09/2020' '20/05/2021' '12/11/2019' '26/10/2021' '08/10/2022'
 '16/03/2020' '22/06/2022' '26/01/2022' '29/07/2021' '25/05/2024'
 '20/06/2020' '24/04/2022' '01/07/2019' '13/11/2023' '13/05/2021'
 '06/09/2021' '09/01/2021' '20/10/2022' '09/08/2021' '04/09/2020'
 '21/11/2019' '25/10/2021' '12/06/2020' '30/03/2020' '12/02/2023'
 '18/03/2022' '01/10/2021' '15/09/2022' '01/05/2022' '06/08/2019'
 '13/01/2023' '18/01/2024' '15/11/2019' '08/09/2021' '24/11/2020'
 '07/01/2021' '03/10/2020' '19/10/2019' '15/05/2020' '22/03/2023'
 '29/01/2021' '28/11/2020' '03/06/2023' '26/07/2023' '23/10/2023'
 '16/01/2020' '03/12/2023' '11/10/2020' '18/09/2021' '14/06/2023'
 '10/03/2020' '12/03/2022' '23/0

## **Standardizing Data Types and Handling Missing Values in `df1`**

### **Converting 'SignupDate' to DateTime**

This code is used to convert a 'SignupDate' column in a pandas DataFrame to a proper `datetime` format while handling different formats and invalid dates.

1. Identify Specific Date Format (`dd/mm/yyyy`):
   - The code first checks which rows follow the `dd/mm/yyyy` format using a regular expression.
   
2. Convert to DateTime:
   - For rows that match this format, it converts the dates into proper `datetime` format. If the format doesn't match, it sets the value to `NaT` (Not a Time).

3. Handle Remaining Dates:
   - The rest of the 'SignupDate' column is also converted to `datetime`, replacing any invalid dates with `NaT`.

4. Check for Invalid Dates:
   - The code checks how many dates could not be converted (i.e., remain as `NaT`) and prints the count.

5. Check Data Type:
   - It prints the data type of 'SignupDate' to verify that it has been converted to `datetime`.

This ensures that all date formats are handled properly, and any invalid dates are tracked.


In [17]:
# Convert 'SignupDate' to datetime, handling different formats
invalid_format = df1['SignupDate'].astype(str).str.contains(r'\d{2}/\d{2}/\d{4}', na=False)
df1.loc[invalid_format, 'SignupDate'] = pd.to_datetime(df1.loc[invalid_format, 'SignupDate'], format='%d/%m/%Y', errors='coerce')

# Convert all 'SignupDate' entries to datetime, coercing errors
df1['SignupDate'] = pd.to_datetime(df1['SignupDate'], errors='coerce')

# Retain only the date part, removing the time component
df1['SignupDate'] = df1['SignupDate'].dt.date

# Check for invalid dates
invalid_dates = df1['SignupDate'].isnull()
num_invalid_dates = invalid_dates.sum()

# Print results
print(f"Number of invalid dates: {num_invalid_dates}")
print(f"Data type of 'SignupDate': {type(df1['SignupDate'].iloc[0]) if not df1['SignupDate'].isnull().all() else 'All NaN'}")

Number of invalid dates: 0
Data type of 'SignupDate': <class 'datetime.date'>


### **Converting remaining `df1` columns to the correct data type**

This code snippet focuses on changing the data types of several columns in a pandas DataFrame (`df1`). Here's a breakdown of each step:

1. Changing Data Types of Specific Columns:
   - The `CustomerID` column is converted to a `string` data type.
   - The `Gender` and `IncomeLevel` columns are converted to `category` type. Categories are useful for columns with a limited set of possible values, optimizing memory usage.
   - The `SignupDate` column is converted to `datetime` format, specifying the correct date format (`dd/mm/yyyy`).

2. Handling Numeric Data with 'Age':
   - The `Age` column is converted to a numeric type, with `errors='coerce'` ensuring that any invalid values are replaced with `NaN` (Not a Number), which represents missing or invalid data.

3. Verifying the Changes:
   - Finally, the data types of all columns in the DataFrame are printed to confirm that the changes were applied correctly.

This ensures that the DataFrame uses the appropriate data types for each column, which can improve both memory efficiency and the correctness of operations on the data.


In [21]:
# Changing data types
df1['CustomerID'] = df1['CustomerID'].astype('string')  # Change to string
df1['Location'] = df1['Location'].astype('string')  # Change to string
df1['CustomerID'] = df1['CustomerID'].astype('string')  # Change to string
df1['Gender'] = df1['Gender'].astype('category')  # Change to category
df1['IncomeLevel'] = df1['IncomeLevel'].astype('category')  # Change to category

# Convert 'SignupDate' to datetime, handling the specific format
df1['SignupDate'] = pd.to_datetime(df1['SignupDate'], format='%d/%m/%Y', errors='coerce')  # Change to datetime, specifying the correct format

# Retain only the date part, removing the time component
df1['SignupDate'] = df1['SignupDate'].dt.date

# Convert 'Age' to numeric, replacing invalid values with NaN
#df1['Age'] = df1['Age'].astype(int) # The original code was causing an error due to trying to convert a string with a decimal to an integer.
df1['Age'] = pd.to_numeric(df1['Age'], errors='coerce').astype('Int64') # Use pd.to_numeric to handle potential errors and convert the column to a numeric type. Then convert to Int64 to handle missing values.

# Verify the changes
print(df1.dtypes)

CustomerID     string[python]
Age                     Int64
Gender               category
Location       string[python]
IncomeLevel          category
SignupDate             object
dtype: object


*Note*

The `SignupDate` column shows as `object` because we converted it to a date type using `.dt.date`. In Python, the date type does not retain time information, resulting in pandas treating it as an object type.

It's important to note that converting a date to a `datetime64` type in pandas will automatically include a time component, defaulting to midnight (00:00:00). This means that if you need to maintain only the date without the time, you'll have to work with the date type or choose to format your datetime as needed during analysis.


### **Handling Missing Values in `df1`**

Handling missing values is vital for maintaining data quality. In this section, we address gaps in the Age and IncomeLevel columns of the DataFrame df1.

We calculate the mean age to fill in missing entries, ensuring a representative value, and introduce the "Uncategorized" category for any missing income levels. This systematic approach enhances the integrity of our dataset, preparing it for more reliable analysis.

### Cleaning the `Age` Column

In this analysis, we focus on cleaning and validating the Age data within our dataset, df1.

During the initial exploration of the dataset, we identified an anomalous age value of 150. This entry is considerably outside the expected range for human ages, suggesting it may be an outlier or a data entry error. To maintain the integrity of our analysis, we decided to implement a systematic approach to handle such outliers and improve the overall quality of the age data.

Our cleaning process involves the following steps:

1. Identifying and removing outliers based on a defined age range (0 to 120 years).
2. Calculating the mean age from the cleaned data.
3. Filling any missing values in the Age column with the calculated mean to ensure a complete dataset.

In [22]:
# Step 1: Identify and handle outliers
min_age = 0
max_age = 120
outliers = df1[(df1['Age'] < min_age) | (df1['Age'] > max_age)]

# Remove outliers
df1 = df1[(df1['Age'] >= min_age) & (df1['Age'] <= max_age)]

# Step 2: Calculate the mean age after outlier removal
mean_age = df1['Age'].mean()

# Step 3: Impute missing values with the mean
df1['Age'].fillna(mean_age, inplace=True)

# Step 4: Recheck for any remaining issues
age_description = df1['Age'].describe()

# Step 5: Print the results with better formatting
print("\n=== Age Statistics After Cleaning ===")
print(f"Count: {age_description['count']}")
print(f"Mean: {age_description['mean']:.2f}")
print(f"Standard Deviation: {age_description['std']:.2f}")
print(f"Minimum: {age_description['min']}")
print(f"25th Percentile: {age_description['25%']}")
print(f"Median (50th Percentile): {age_description['50%']}")
print(f"75th Percentile: {age_description['75%']}")
print(f"Maximum: {age_description['max']}")
print("=====================================\n")



=== Age Statistics After Cleaning ===
Count: 2650.0
Mean: 44.64
Standard Deviation: 15.41
Minimum: 18.0
25th Percentile: 31.0
Median (50th Percentile): 45.0
75th Percentile: 58.0
Maximum: 70.0



<ipython-input-22-0f46e5c2e365>:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['Age'].fillna(mean_age, inplace=True)


In [23]:
# Add "Uncategorized" as a new category to the 'IncomeLevel' column
df1['IncomeLevel'] = df1['IncomeLevel'].cat.add_categories('Uncategorized')

# Fill missing values in the 'IncomeLevel' column with 'Uncategorized'
df1['IncomeLevel'].fillna('Uncategorized', inplace=True)

# Check if there are any remaining missing values in 'IncomeLevel'
missing_income = df1['IncomeLevel'].isnull().sum()

# Print the results
print(f"Missing values in 'IncomeLevel': {missing_income}")


Missing values in 'IncomeLevel': 0


<ipython-input-23-e4407e128bb7>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['IncomeLevel'] = df1['IncomeLevel'].cat.add_categories('Uncategorized')
<ipython-input-23-e4407e128bb7>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['IncomeLevel'].fillna('Uncategorized', inplace=True)


## **Standardizing Data Types and Handling Missing Values in `df2`**

### **Converting 'TransactionDate' to DateTime**

In [24]:
# Convert 'TransactionDate' to datetime, handling different formats
# First, identify rows with valid date format
valid_format = df2['TransactionDate'].astype(str).str.contains(r'\d{2}/\d{2}/\d{4}', na=False)  # Convert 'TransactionDate' to string type temporarily

# Convert valid date formats to datetime
df2.loc[valid_format, 'TransactionDate'] = pd.to_datetime(df2.loc[valid_format, 'TransactionDate'], format='%d/%m/%Y', errors='coerce')

# Convert all 'TransactionDate' entries to datetime, coercing errors
df2['TransactionDate'] = pd.to_datetime(df2['TransactionDate'], errors='coerce')

# Retain only the date part, removing the time component
df2['TransactionDate'] = df2['TransactionDate'].dt.date

# Check for invalid dates
invalid_dates = df2['TransactionDate'].isnull()
num_invalid_dates = invalid_dates.sum()

# Print results
print(f"Number of invalid dates: {num_invalid_dates}")
print(f"Data type of 'TransactionDate': {type(df2['TransactionDate'].iloc[0]) if not df2['TransactionDate'].isnull().all() else 'All NaN'}")


Number of invalid dates: 0
Data type of 'TransactionDate': <class 'datetime.date'>


### **Converting remaining `df2` columns to the correct data type**

In [26]:
df2['CustomerID'] = df2['CustomerID'].astype('string')
df2['TransactionID'] = df2['TransactionID'].astype('string')

df2['ProductCategory'] = df2['ProductCategory'].astype('category')
df2['PaymentMethod'] = df2['PaymentMethod'].astype('category')

# Convert 'Amount' to float, replacing invalid values with NaN
df2['Amount'] = pd.to_numeric(df2['Amount'], errors='coerce')

# Round 'Amount' to 2 decimal points
df2['Amount'] = df2['Amount'].round(2)

# Verify the changes
print(df2.dtypes)

CustomerID         string[python]
TransactionID      string[python]
TransactionDate            object
Amount                    float64
ProductCategory          category
PaymentMethod            category
dtype: object


### **Handling Missing values in `df2`**

In [41]:
# Calculate the mean of the 'Amount' column, excluding NaN values
mean_amount = df2['Amount'].mean()

# Fill missing values in the 'Amount' column with the calculated mean
df2['Amount'].fillna(mean_amount, inplace=True)

# Round the 'Amount' column to 2 decimal points
df2['Amount'] = df2['Amount'].round(2)

# Add "Uncategorized" as a new category to the 'ProductCategory' column if it doesn't exist
if 'Uncategorized' not in df2['ProductCategory'].cat.categories:
    df2['ProductCategory'] = df2['ProductCategory'].cat.add_categories('Uncategorized')

# Fill missing values in the 'ProductCategory' column with 'Uncategorized'
df2['ProductCategory'].fillna('Uncategorized', inplace=True)

# Check if there are any remaining missing values in 'Amount' and 'ProductCategory'
missing_amount = df2['Amount'].isnull().sum()
missing_product = df2['ProductCategory'].isnull().sum()

# Print the results with mean formatted to 2 decimal points
print(f"Mean Amount (filled): {mean_amount:.2f}")
print(f"Missing values in 'Amount': {missing_amount}")
print(f"Missing values in 'ProductCategory': {missing_product}")


Mean Amount (filled): 489.79
Missing values in 'Amount': 0
Missing values in 'ProductCategory': 0


## **Standardizing Data Types and Handling Missing Values in `df3`**

### **Converting 'InteractionDate' to DateTime**

In [29]:
# Convert 'InteractionDate' to datetime, handling different formats
# First, identify rows with valid date format
valid_format = df3['InteractionDate'].astype(str).str.contains(r'\d{2}/\d{2}/\d{4}', na=False)  # Convert 'InteractionDate' to string type temporarily

# Convert valid date formats to datetime
df3.loc[valid_format, 'InteractionDate'] = pd.to_datetime(df3.loc[valid_format, 'InteractionDate'], format='%d/%m/%Y', errors='coerce')

# Convert all 'InteractionDate' entries to datetime, coercing errors
df3['InteractionDate'] = pd.to_datetime(df3['InteractionDate'], errors='coerce')

# Check for invalid dates
invalid_dates = df3['InteractionDate'].isnull()
num_invalid_dates = invalid_dates.sum()

# Print results
print(f"Number of invalid dates: {num_invalid_dates}")
print(f"Data type of 'InteractionDate': {df3['InteractionDate'].dtype}")

Number of invalid dates: 0
Data type of 'InteractionDate': datetime64[ns]


### **Converting remaining `df3` columns to the correct data type**

In [30]:
df3['CustomerID'] = df3['CustomerID'].astype('string')
df3['InteractionID'] = df3['InteractionID'].astype('string')
df3['InteractionDate'] = df3['InteractionDate'].dt.date

df3['Platform'] = df3['Platform'].astype('category')
df3['InteractionType'] = df3['InteractionType'].astype('category')
df3['Sentiment'] = df3['Sentiment'].astype('category')

# Verify the changes
print(df3.dtypes)

CustomerID         string[python]
InteractionID      string[python]
InteractionDate            object
Platform                 category
InteractionType          category
Sentiment                category
dtype: object


### **Handling Missing values in `df3`**

In [31]:
# Add "Uncategorized" as a new category to the 'Platform' column if it doesn't exist
if 'Uncategorized' not in df3['Platform'].cat.categories:
    df3['Platform'] = df3['Platform'].cat.add_categories('Uncategorized')

# Fill missing values in the 'Platform' column with 'Uncategorized'
df3['Platform'].fillna('Uncategorized', inplace=True)

# Add "Uncategorized" as a new category to the 'Sentiment' column if it doesn't exist
if 'Uncategorized' not in df3['Sentiment'].cat.categories:
    df3['Sentiment'] = df3['Sentiment'].cat.add_categories('Uncategorized')

# Fill missing values in the 'Sentiment' column with 'Uncategorized'
df3['Sentiment'].fillna('Uncategorized', inplace=True)

# Check if there are any remaining missing values in 'Platform' and 'Sentiment'
missing_platform = df3['Platform'].isnull().sum()
missing_sentiment = df3['Sentiment'].isnull().sum()

# Print the results
print(f"Missing values in 'Platform': {missing_platform}")
print(f"Missing values in 'Sentiment': {missing_sentiment}")

Missing values in 'Platform': 0
Missing values in 'Sentiment': 0


## **Checking Data for remaining duplicates and missing values**

In [32]:
# Loop through each DataFrame to check for duplicates
print("=== Checking for Duplicates ===")
for i, df in enumerate([df1, df2, df3], 1):
    num_duplicates = df.duplicated().sum()
    print(f"DataFrame {i}: Number of duplicates = {num_duplicates}")

print("\n" + "="*50 + "\n")  # Separator for readability

# Loop through each DataFrame to check for missing values
print("=== Checking for Missing Values ===")
for i, df in enumerate([df1, df2, df3], 1):
    print(f"DataFrame {i}:")

    # Calculate missing values for each column
    missing_values = df.isnull().sum()

    # Print columns with missing values
    missing_cols = missing_values[missing_values > 0]
    if not missing_cols.empty:
        print("Missing values found in the following columns:")
        print(missing_cols)
    else:
        print("No missing values found.")

    print("\n" + "="*50 + "\n")  # Separator for readability

# Loop through each DataFrame to check data types
print("=== Checking Data Types ===")
for i, df in enumerate([df1, df2, df3], 1):
    print(f"DataFrame {i}:")
    # Print data types of each column
    print(df.dtypes)
    print("\n" + "="*50 + "\n")  # Separator for readability


=== Checking for Duplicates ===
DataFrame 1: Number of duplicates = 0
DataFrame 2: Number of duplicates = 0
DataFrame 3: Number of duplicates = 0


=== Checking for Missing Values ===
DataFrame 1:
No missing values found.


DataFrame 2:
No missing values found.


DataFrame 3:
No missing values found.


=== Checking Data Types ===
DataFrame 1:
CustomerID     string[python]
Age                     Int64
Gender               category
Location       string[python]
IncomeLevel          category
SignupDate             object
dtype: object


DataFrame 2:
CustomerID         string[python]
TransactionID      string[python]
TransactionDate            object
Amount                    float64
ProductCategory          category
PaymentMethod            category
dtype: object


DataFrame 3:
CustomerID         string[python]
InteractionID      string[python]
InteractionDate            object
Platform                 category
InteractionType          category
Sentiment                category
dtype: ob

## **Exporting Cleaned Data**

In [43]:
df1.to_csv('/content/drive/MyDrive/3rd year 2024-25/MO-IT125 - Math for Machine Learning: Linear Algebra/cleaned/df1.csv', index=False)
df2.to_csv('/content/drive/MyDrive/3rd year 2024-25/MO-IT125 - Math for Machine Learning: Linear Algebra/cleaned/df2.csv', index=False)
df3.to_csv('/content/drive/MyDrive/3rd year 2024-25/MO-IT125 - Math for Machine Learning: Linear Algebra/cleaned/df3.csv', index=False)

## **Method for Handling Missing Values**

A. For any numeric value such as Age and Amount, calculate the mean value and populate the missing data with the mean value. The mean value is used instead of leaving these blank.

B. For each record with missing data in some string columns, create a new label called Uncategorized and apply these to the said columns.

1. Income Level
2. Product Category
3. Platform
4. Sentiment